### PR TITLE
Fix slice selector large n

### LINKED ISSUE #153: [BUG] Slice selector breaks with a lot of slices
Issue Description
-----------------

Trying to explore a computation with 10k slices. The selector bereaks.

How to Reproduce
----------------

Upload a big JSON with 10k lines or use a script to duplicate slices until you have 10k.

### Error Output

image
5k pixels to the right:
image

### PR DESCRIPTION
**Pull request by Keelando**
Fix slice selector to accept wide range of inputs, 1 <= N <= 10k. 

Should theoretically work with arbitrary length of slices

closes #153

### CODE REVIEW

**Comment by Keelando** [comment]
Should we leave the slice index counting from zero or from one?

**Comment by gwwatkin** [comment]
> Should we leave the slice index counting from zero or from one?

I'm in favour of zero.

File: src/components/SliceIndexBar.tsx [hunk 1]
```diff
@@ -1,39 +1,75 @@
...
+    }, [selected])
     return (
-        <Box borderWidth="3px" borderColor="black" maxW="400px" rounded="xl" p="1" boxShadow={"lg"}>
-            <Flex gap={flex_gap}>
-                {nElements.map(function (element, index) {
-                    const color = index <= selected ? "#4299e1" : "lightgrey"
+        <Box borderWidth="3px" borderColor="black" maxW="450px" rounded="lg" p="1" boxShadow="md">
+            <Flex>
+                {nSteps.map((index) => {
+                    const color = (index * count) / max_steps <= selected ? "#4299e1" : "lightgrey"
```
Comment (gwwatkin) [inline src/components/SliceIndexBar.tsx:43]:
```suggestion
                    const color = (index * count) / max_steps <= selected ? "#4299e1" : "#edf2f7"
```

**Review by gwwatkin — APPROVED** [review]
Fantastic, I can only add a suggestion about the colour, to make it consistent with the rest of the page.

**Comment by gwwatkin** [comment]
Also, maybe can you get the slider to start from 0? (so that in the first slice it's all gray)

I see why you would want to start at 1 in this case.


One more thought is, why don't we just make the slider a little wider to increase the resolution? does it relate to the mobile view? anyways not important

**Comment by Keelando** [comment]
I added a max_width parameter that lets us change the desired width. It will be tricky to make mobile friendly because once you render screens <500px wide, you need to subdivide pixels (which doesn't work) to get the resolution you want. The solution is to reduce the sensitivity of the scale. That is why I initially chose a resolution of 100 steps (since changed to 200). 

> Also, maybe can you get the slider to start from 0? (so that in the first slice it's all gray)

I found it confusing to use the slice selector for small N, say 0<9 when that behaviour is enabled.

In the case where there are 4 slices, it makes sense to subdivide the bar into 4 pieces, but if you want the slice to be grey, that means when you click the 4th slice, you only get 3 blue rectangles shaded out of 4. However, when N is large, you don't notice that edge case at all. If you want, I can enable your desired functionality for large N.

File: src/components/SliceIndexBar.tsx [hunk 2]
```diff
@@ -1,40 +1,83 @@
...
+            borderWidth="3px"
+            borderColor="black"
+            maxW={`${max_width}px`}
+            rounded="lg"
+            p="1"
+            boxShadow="md"
+        >
+            <Flex>
+                {nSteps.map((index) => {
+                    const color = (index * count) / max_steps <= selected ? "blue.300" : "lightgrey"
```
Comment (gwwatkin) [inline src/components/SliceIndexBar.tsx:55]:
```suggestion
                    const color = (index * count) / max_steps <= selected ? "blue.300" : "edf2f7"
```
It as the gray I was changing

**Review by gwwatkin — APPROVED** [review]
> @gwwatkin I added a max_width parameter that lets us change the desired width. It will be tricky to make mobile friendly because once you render screens <500px wide, you need to subdivide pixels (which doesn't work) to get the resolution you want. The solution is to reduce the sensitivity of the scale. That is why I initially chose a resolution of 100 steps (since changed to 200).

Ok np. Lets make it high resolution desktop. If we ever have mobile users, I think it's ok if they have to scroll a little to the side, though I think you can put an if on view port width.

> >  Also, maybe can you get the slider to start from 0? (so that in the first slice it's all gray)

> I found it confusing to use the slice selector for small N, say 0<9 when that behaviour is enabled.
> In the case where there are 4 slices, it makes sense to subdivide the bar into 4 pieces, but if you want the first slice to be grey when slice 0 is checked out, that means when you click the 4th slice, you only get 3 blue rectangles shaded out of 4. However, when N is large, you don't notice that edge case at all. If you want, I can enable your desired functionality for large N, though that might involve some more weird edge case logic to have different behaviour at small/large N

I'm fine with how it is, but if you really wanted to make it perfect you could make the the number of ticks N-1 maybe? so that you have both cases covered. I encourage you to merge though. :shipit:

File: src/components/SliceIndexBar.tsx (hunk 2 above)
Comment (Keelando) [reply src/components/SliceIndexBar.tsx:55]: Oh yeah I see now.

**Comment by Keelando** [comment]
N-1 ticks means you lost the ability to click the first one. I'll merge and as we use it, we can modify it again.

**Comment by gwwatkin** [comment]
> N-1 ticks means you lost the ability to click the first one. I'll merge and as we use it, we can modify it again.

Yeah didn't think of that :thinking:

**Pull request merged**